### PR TITLE
factored out persistence REST resources into own bundle

### PR DESCRIPTION
**Pull request by sjka**
I've been running into weird problems with the REST resources inside the core.persistence bundle. Whenever io.rest gets installed _after_ persistence.core, then DS attempts to activate the persistence rest resource component, failing badly as the RestResource base class cannot be found. 

When the io.rest bundle gets installer later on, this issue does not fix itself, as the OSGi spec does not require any existing bundles to be re-resolved only to satisfy _optional_ dependencies.

So although I know it's perfectly legal as it is, I would still suggest to factor out the rest dependencies from the core for the above reasons.

Signed-off-by: Simon Kaufmann <simon.kfm@googlemail.com>

### CODE REVIEW

**Comment by maggu2810** [comment]
Have not reviewed your changes until now, but your summary sounds good to me.

> as the OSGi spec does not require any existing bundles to be re-resolved only to satisfy optional dependencies

That the reason why I don't like optional dependencies and does not use it in any of my projects (where the decision is up to me).

File: bundles/io/org.eclipse.smarthome.io.rest.persistence/META-INF/MANIFEST.MF [hunk 1]
```diff
@@ -0,0 +1,18 @@
+Manifest-Version: 1.0
+Bundle-ManifestVersion: 2
+Bundle-Name: Eclipse SmartHome Core Persistence Rest
+Bundle-SymbolicName: org.eclipse.smarthome.io.rest.persistence
+Bundle-Version: 0.8.0.qualifier
+Bundle-Vendor: Eclipse.org/SmartHome
+Bundle-RequiredExecutionEnvironment: JavaSE-1.7
+Import-Package: org.eclipse.smarthome.core.items,
```
Comment (maggu2810) [inline bundles/io/org.eclipse.smarthome.io.rest.persistence/META-INF/MANIFEST.MF]: Can you use alphabetical order for all imports?
That is not a must, but could improve readability.

File: bundles/core/org.eclipse.smarthome.core.persistence/META-INF/MANIFEST.MF [hunk 2]
```diff
@@ -6,14 +6,10 @@ Bundle-Version: 0.8.0.qualifier
...
- javax.ws.rs.core;resolution:=optional,
- org.eclipse.smarthome.core.items,
+Import-Package: org.eclipse.smarthome.core.items,
  org.eclipse.smarthome.core.library.types,
  org.eclipse.smarthome.core.persistence,
  org.eclipse.smarthome.core.types,
- org.eclipse.smarthome.io.rest;resolution:=optional,
  org.slf4j
 Export-Package: org.eclipse.smarthome.core.persistence
 Service-Component: OSGI-INF/*.xml
```
Comment (maggu2810) [inline bundles/core/org.eclipse.smarthome.core.persistence/META-INF/MANIFEST.MF]: You should remove that line if it does not exist in the bundle.

Comment (sjka) [inline bundles/core/org.eclipse.smarthome.core.persistence/META-INF/MANIFEST.MF]: ouch, yes, of course!

File: bundles/io/org.eclipse.smarthome.io.rest.persistence/META-INF/MANIFEST.MF (hunk 1 above)
Comment (sjka) [inline bundles/io/org.eclipse.smarthome.io.rest.persistence/META-INF/MANIFEST.MF]: yes, good point

**Comment by sjka** [comment]
> That the reason why I don't like optional dependencies and does not use it in any of my projects (where the decision is up to me).

good to know I'm not alone here ;-)

Thanks for the quick feedback, I have updated the PR

**Comment by kaikreuzer** [comment]
If we decide not to use optional dependencies for components, could you please scan the COMPLETE project and see how many new bundles we have to introduce?
If we take this decision it should imho be applied to all occurences at the same time.

**Comment by kaikreuzer** [comment]
Note that there have been other discussions on keeping the number of bundles to a minimum (I remember @dvanherbergen being strong about that point).

**Comment by maggu2810** [comment]
Do we need the all or nothing decision here or could we merge this one now?

IMHO we should create good OSGi bundles that work independent of the time they are resolved.
As ESH is a framework that is used in different products it is not possible to fit all needs (keep the code size as small as possible as using just bundles / code /functions that are needed or use big bundles that contains as lot as code as possible to keep the number of bundles small).

A quick look at google doesn't give me a result if the number of bundles have a big performance impact. IIRC the number of bundles comments of @dvanherbergen has been about the Whiteboard feature that scans every bundle and so increase the startup time. But this is nothing that is caused by ESH.

**Comment by maggu2810** [comment]
@sjka You have missed that file: https://github.com/eclipse/smarthome/blob/master/features/karaf/src/main/feature/feature.xml

**Comment by sjka** [comment]
oh, that indeed was not on my radar. Sorry!

**Comment by dvanherbergen** [comment]
@maggu2810, that is correct. PAX Jetty scans all the bundles at startup for [Java services](https://docs.oracle.com/javase/tutorial/ext/basics/spi.html) (not OSGI services). During this scan it unzips and reads the content of every bundle. When running on a device with limited CPU and disk performance like a pi, the impact is very noticeable. When running on a 'regular' machine, this impact is not noticeable. For the PAX Jetty case, that is why we have the patched Jetty jar. This removes the scanning part as it is not required for openHAB.

Another example could be the eclipse IDE. If you start it with the -clean option, startup performance drops significantly. But I'm not sure if that is caused by the number of bundles or the number of services.

Besides performance, I personally prefer a lower amount of bundles because I believe it makes things easier to understand. In my opinion, sometimes the bundles use a granularity that is too low, e.g. the automation bundles. There I would have expected 2 bundles. An API and an implementation. 
In the implementation bundle, you could use the approach that all included services are loaded by default, but can be replaced with a service implementation in an external bundle.
This way, you keep the packaging simple. If you want to use the default implementation for everything, you just use the two bundles. If you want to change the automation completely with your custom implementation, you just use the API. If you want to change only a single service, you use the two bundles and add your own service in a third which would override the default service. 
But this is all very subjective and a design decision, i.e. do you focus your design on the 5% of users that would want to mix and match fine grained bundles, or do you focus your design on the 95% of users who will just use the default stuff and would be better served with less bundles.

Keeping REST code separated from the rest seems like a good idea to me, as I can imagine a few cases, where you would want to be able to run the application without REST.

**Comment by maggu2810** [comment]
@dvanherbergen Thank you for your response. Is it okay for you (and @kaikreuzer and @sjka) to move this discussion about bundle count to some other place (e.g. forum, mailing list, separate issue)?
I will create a cross-link here.

If it is okay for you (at least @kaikreuzer) to split the REST? I would like to merge this one.

**Comment by dvanherbergen** [comment]
Sounds good to me.

**Comment by kaikreuzer** [comment]
> If it is okay for you (at least @kaikreuzer) to split the REST?

Yes, although two questions remain for me:
1. What @sjka describes as the primary reason for this PR sounds like a critical bug to me - so as mentioned above, we should make sure that if this is the case, we fix potential other occurrences as well (from my quick check, persistence-rest is the only one, so we should be fine - but a second pair of eyes to check that would be good :-))
2. As we are moving it away from core.persistence - wouldn't the correct place to move it to be the (already existing) io.rest.core bundle? In that case, we would not even increase the number of bundles.

**Comment by sjka** [comment]
1.) Yes, indeed. So far I observed this bug only in relation to REST, but that might be coincidence. Nevertheless, you are right, we should fix it _at least_ for all occurrences. My quick search came up with those two additional bundles:

```
find . -name "MANIFEST.MF" -exec grep "rest;resolution:=optional" /dev/null {} \;                                                                                                
./bundles/core/org.eclipse.smarthome.core.id/META-INF/MANIFEST.MF: org.eclipse.smarthome.io.rest;resolution:=optional,
./bundles/ui/org.eclipse.smarthome.ui.icon/META-INF/MANIFEST.MF: org.eclipse.smarthome.io.rest;resolution:=optional,
```

2.) Regarding the correct place: to my understanding core.persistence is kind of an optional bundle. If we now add it to io.rest.core, we introduce a mandatory dependency to core.persistence, meaning that whenever you want to use REST, you will have to have persistence too. Not the idea, right? 
(We _could_ mark it as optional there, but would end up in the same mess then, just the other way round, so that's not really feasible...)

**Comment by cdjackson** [comment]
> What @sjka describes as the primary reason for this PR sounds like a critical bug to me

I would agree. I've not personally seen this, but if this is correct, then it effectively means that to be sure that REST services work, they must be installed in the correct order and in this case moving the persistence REST resources into the rest.core bundle isn't really a "solution" - it's just a workaround for this particular resource. 

However, for other resources that might be provided in a separate bundle, it doesn't work. Surely the solution should be to ensure that REST resources registering before the core bundle are still able to work? I don't know if this is possible, but it seems like the correct way to actually solve this issue?

**Comment by maggu2810** [comment]
@cdjackson The order does matter if we are using optional resolution. We should drop the optional resolution and to go this way, the split makes sense.
It is not the OSGi way to rely on the order.

**Comment by maggu2810** [comment]
I think I agree with @sjka that as long as persistence is not a mandatory component we should use a separate bundle.
So, I am waiting for the next "ready to merge" :wink:

**Comment by cdjackson** [comment]
> It is not the OSGi way to rely on the order.

Sure - I understand that, that's why I said we should make it work if the rest resource registered/started before the rest server - or am I wrong?

If we go down this route, then aren't we effectively requiring all rest resources to be in the same bundle - it just seems wrong.

I'm not an expert on OSGi so I'm sure you know better than me, but it just seems to me that we a) shouldn't rely on the order, and b) shouldn't need to move the rest resources into the same bundle to resolve the problem.

**Comment by maggu2810** [comment]
@cdjackson Perhaps I misunderstood your comment (or you mine) :wink:  
If I understand your last post correctly, I agree with a) and b) (before and now).

**Comment by kaikreuzer** [comment]
> My quick search came up with those two additional bundles [...]

Ok, so shall we create a new separat issue for this (or even two issues?)? I would like to clarify the question (2) on WHERE to move them definitely already now.

> as long as persistence is not a mandatory component we should use a separate bundle.

Well, I would claim it is - there really is nothing different to core.binding or core.thing - they all are designed to be "optional", but for pragmatic reasons (concretely: for not having too many bundles), we at the time decided to put all core.xxx rest resources in a single bundle.

Apart from this, I really expect persistence to be a "real" core component in future once we go for the new sitemap concept as well as adding functionality to the new rule engine. Treating it as optional would mean that we would have to separate out all persistence-related functionality in those new features into separate bundles as well - I really think that this will cause us problems. So unless there is any strong reason why it cannot be added to rest.core as well, I would opt for that location.

**Comment by maggu2810** [comment]
If we state that persistence is not optional anymore, it is IMHO then okay to move it to io.rest.core.

But if we start moving all non-optional bundles together, what about other bundles?
I assume we must then also move io.rest.sitemap to io.rest, correct?
Have we to move core.thing, core.persistence, core.library to core, if them are not really optional?
Or do we still want to keep some functional groups (not only by using package names but also bundles)? I am fine with keeping other stuff separated and move all rest parts together, but if we argue to use one bundle for io.rest.core and io.rest.core.persistence I need to know what differs from core and core.persistence.

Can someone point me to some documents or some information that the number of bundles does really matter? If it is just related to some third party code that scans all bundles, I don't think that it is a general problem then. But I started never a measurement myself. I am happy if someone could point me to some informations about that topic.

So, some options (you can choose multiple ones):
* using io.rest.core, io.rest.sitemap, io.rest.see and io.rest.persistence
* using io.rest that contians all rest related resources
* using io.rest.core (with core and persistence) and keep io.rest.sitemap and io.rest.see (why?)
* create separate issues to merge other bundles
* create separate issues to remove other optional resulutions, too

**Comment by kaikreuzer** [comment]
> we must then also move io.rest.sitemap to io.rest, correct?

No, sitemap is not core.sitemap, it is model.sitemap and hence we could rather rename it to model.sitemap.rest (in analogy to automation.rest). Note that this has been deliberately kept separately not to introduce any dependency on EMF/Xtext stuff in the core.

> Have we to move core.thing, core.persistence, core.library to core

Note that core.library already HAS moved to core.

As I said, the rest resources were put into a common bundle for pragmatic reasons - nobody really wanted to create so many different bundles with each just having a single class inside (and that was rather about manual code maintenance than about runtime performance). The idea is that if you want to have some stripped down ESH-node, you are likely to not have REST and hence you are still able to decide that you might not want to have thing/binding/persistence support either, but e.g. only run a minimal event bus. So having the core split, while keeping the rest-bundle combined makes sense.

**Comment by maggu2810** [comment]
> No, sitemap is not core.sitemap, it is model.sitemap and hence we could rather rename it to model.sitemap.rest (in analogy to automation.rest). Note that this has been deliberately kept separately not to introduce any dependency on EMF/Xtext stuff in the core.

Sound plausibly, thanks for clarification.
 
> Note that core.library already HAS moved to core.

Then it would be nice to get rid of the directories (#1372).

@sjka If persistence is not optional, are you fine to move the code to io.rest.core?

**Comment by sjka** [comment]
sure, I will update this PR accordingly

**Comment by dvanherbergen** [comment]
Sorry for dragging the discussion out, but I just thought of another option: What if the bundles wouldn't be changed, but the RestResource class would be moved to the core bundle instead? Would that work or are there other dependencies that would cause an issue?
The RestResource class could be considered to be part of the core API and including it there would not prevent running a setup without the REST implementation.

**Comment by maggu2810** [comment]
@dvanherbergen What is wrong about moving the REST stuff into one bundle? What is the advantage of moving RestResource to the core bundle?

**Comment by dvanherbergen** [comment]
There is nothing wrong with moving the REST stuff into one bundle. I just wanted to highlight that there was another option.
The main difference would be that you could keep the optional dependency of the persistence bundle on the rest bundle.

**Comment by maggu2810** [comment]
But at least @sjka and @maggu2810 are happy to drop them. :wink: 
In this issue I see no comment "we do not want to drop the optional resolution because of ...", so I assume no one really wants to keep them.

**Comment by kaikreuzer** [comment]
I think @dvanherbergen point is to only move the interface RESTResource. By doing so, we would not have to adapt core.id and ui.icon either and could happily in future include rest resources in new feature bundles without requiring to reopen this discussion (where should the rest resource for the new feature xy go?). I think this suggestion has some charme, doesn't it?

**Comment by maggu2810** [comment]
The error is caused by the missing re-resolve of optional dependencies.
It could be masked by moving file to other bundles or it could be solved by dropping optional dependencies.
Moving the file does only mask the wired behavior.
I assume the intention of this issue has not been to drop optional resolution for the whole project but in that places where it introduce a failure.
I don't think it is good to keep "optional" and just moving stuff around to solve it.

**Comment by kaikreuzer** [comment]
@maggu2810 I think you need to differentiate between
javax.ws.rs;resolution:=optional
and
org.eclipse.smarthome.io.rest;resolution:=optional

As far as I understand, the problem that @sjka describes is caused by the optional io.rest import only - and this would be solved by @dvanherbergen suggestion.

> But at least @sjka and @maggu2810 are happy to drop them. 

Does this also mean that you both volunteer to create PRs the other occurrences?

**Comment by maggu2810** [comment]
> Does this also mean that you both volunteer to create PRs the other occurrences?

I can volunteer to try to find a solution for the other occurrences of optional resolutions. If I can create a PR depends if am succeed.

Can I merge this one, after @sjka changed this PR to use io.rest.core instead of io.rest.persistence?
(I will still think about the difference, but I assume the wiring order of the bundles still does matter - at least if there is no other bundle that import javax.ws.rs which is a transitive one...)

**Comment by kaikreuzer** [comment]
What speaks against Davys suggestion? If moving the interface solves the issue now and forever in future as well, it sounds like the simplest and best option to me. What am I missing? 
Am 13.04.2016 10:03 vorm. schrieb Markus Rathgeb <notifications@github.com>:
Does this also mean that you both volunteer to create PRs the other occurrences?


I can volunteer to try to find a solution for the other occurrences of optional resolutions. If I can create a PR depends if am succeed.

Can I merge this one, after @sjka changed this PR to use io.rest.core instead of io.rest.persistence?
(I will still think about the difference, but I assume the wiring order of the bundles still does matter - at least if there is no other bundle that import javax.ws.rs which is a transitive one...)

**Comment by maggu2810** [comment]
Okay, both approaches will solve the problem. I like this one because it also drops the optional from the touched manifests.

* RESTResource is only a marker interface but I still don't see why is better to move a "marker interface for REST resource implementations" to core instead of io.rest.core. Sure, it doesn't matter, but I don't know why the one place is better then the other one.
* It seems that there is some reason you don't want to drop that optional stuff (e.g. javax.ws.rs;resolution:=optional) from this bundle. Does this PR break any other ESH stuff?
* I still don't understand what is wrong with this PR (after adaption to use io.rest.core). Perhaps I would understand this long discussion if this PR does something really wrong. For me it looks like we are searching another solution that does involve as few changes as possible.

**Comment by maggu2810** [comment]
Okay, both approaches will solve the problem. I like this one because it also drops the optional from the touched manifests.

RESTResource is only a marker interface but I still don't see why is better to move a "marker interface for REST resource implementations" to core (from io.rest.core) intead of fixing the wiring. Sure, it doesn't matter, but I don't know why the one place is better then the other one.
It seems that there is some reason you don't want to drop that optional stuff (e.g. javax.ws.rs;resolution:=optional) from this bundle. Does this PR break any other ESH stuff?
I still don't understand what is wrong with this PR (after adaption to use io.rest.core). Perhaps I would understand this long discussion if this PR does something really wrong. For me it looks like we are searching another solution that does involve as few changes as possible.

**Comment by maggu2810** [comment]
And what is wrong to move the REST stuff from core.persistence to io.rest.core? Why should we keep the REST code in core.persistence if there is a REST bundle? Isn't that the current approach used by the other REST resources?

**Comment by sjka** [comment]
okay, PR is adapted. 

btw: moving the marker interface RestResource to core doesn't really solve this issue completely: As @maggu2810 correctly pointed out, we will still have the optional imports on javax.ws.rs etc, which potentially are subject to the same flaw.

**Comment by kaikreuzer** [comment]
thanks, all fine for me.

**Pull request merged**